elastic/kibana
# [Synthetics] copy alert state to alert context and implement alert recovery

## Summary

Resolves https://github.com/elastic/kibana/issues/128760
Resolves https://github.com/elastic/kibana/issues/128761

This PR takes the first step towards transitioning Uptime alerting towards preferring context over state. For more information about why this change is important, please visit the ticket.

For more information on the overall stages of the transition, please see https://github.com/elastic/kibana/issues/126280

## Testing that State has been copied over to Context
1. Connect to ES via `oblt-cli`. This ensures you will have at least one monitor that is failing and one with an aging or expired certificate.
2. Navigate to Uptime overview and create a monitor status rule from the alert flyout, making sure to give the rule a name.
3. Select an alert connector. The easiest connector to configure is a server log. Click save.
4. Ensure the alert comes to your server log with the appropriate alert message
5. Navigate to Uptime overview and create a tls rule from the alert flyout, making sure to give the rule a name.
6. Select the server log connector. Click save
7. Ensure the tls alert comes to your server log with the appropriate alert message

**Note**
You'll likely want to set up control rules to ensure that the alert messages match our original implementation and still work for legacy users using state instead of context. 
To do so, create two more rules, one for monitor status and one for tls. Adjust the alert message to the following. Ensure that the control (state) matches the test (context).

Monitor Status
--

```
Monitor {{state.monitorName}} with url {{{state.monitorUrl}}} from {{state.observerLocation}} {{{state.statusMessage}}} The latest error message is {{{state.latestErrorMessage}}}
```

TLS
--

```
Detected TLS certificate {{state.commonName}} from issuer {{state.issuer}} is {{state.status}}. Certificate {{state.summary}}
```

### Testing alert resolution

Monitor Status
--
1. Start your own failing monitor via the synthetics service or running heartbeat from source. This ensures you're able to control your monitor's up and down state. 
2. Create a monitor status alert. To make things simple, I recommend creating an alert for when the monitor has failed more than once within the last 3 minutes, and removing the availability check.
3. Before saving your alert, ensure that you add an action for the Recovery action group. First select a connector type (I like to use Server logs). Then add default action for `Run When: Uptime Monitor Down`. Click `Add action` again. This type change `Run When` to `Recovered`. Confirm that the default content for recovery matches the AC defined in https://github.com/elastic/kibana/issues/128761.
<img width="703" alt="Screen Shot 2022-04-27 at 4 28 08 PM" src="https://user-images.githubusercontent.com/11356435/165625542-905ccebd-40d4-453a-87f7-a616ae7a5a90.png">
4. Wait for the monitor alert to trigger.
5. Force the monitor to resolve by either A.) stopping heartbeat from running or deleting the monitor or B.) Changing the monitor config to a value that would trigger an UP status.
6. Confirm that you receive the alert recovery message to your specified action connector

TLS
--
1. Run your own monitors, or connect to ES via `oblt-cli`.
2. Navigate to Uptime Settings
3. Change age limit for certificates to 1 day
<img width="948" alt="Screen Shot 2022-04-27 at 4 36 40 PM" src="https://user-images.githubusercontent.com/11356435/165629891-4bc4f310-2f44-405f-acc7-23ae331091ed.png">
4. Create a TLS status alert. Make sure you add an action for the when the alert is triggered as well as for when the alert is recovered. Confirm that the default content for recovery matches the AC defined in https://github.com/elastic/kibana/issues/128761.
5. Wait for the alert to trigger
6. Navigate back to Uptime Settings. Change age limit back to 730 days.
7. Confirm that you receive the alert recovery message to your specified action connector
